text/0002-numeric-values.md

```diff
@@ -0,0 +1,485 @@
+- `%`: If the divisor is `0`.
+
+Note that floating point operations cannot overflow, but instead will return `f64.inf` as IEEE 754 specifies.
+
+## NaN
+
+There is no `NaN` constant. Instead, any operation that could produce `NaN` is either safety-checked undefined behavior (like dividing by zero), or will return an optional `f64`. When representing `NaN` is required, it should be represented as an optional `f64`, to indicate the possibility of `NaN`.
+
+If it was missed when writing this RFC. Any math operation that would produce `NaN` should be safety-checked undefined behavior.
+
+> [!NOTE]
+> There is not yet an RFC for optionals.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Parsing numbers
+
+For reference, the following regular expression is a set of all valid numeric literals:
+
+```regexp
+^(?:\d(?:\d|_\d)*)?\.?(?:\d(?:\d|_\d)*)+(?:e[+-]?(?:\d(?:\d|_\d)*))?\b
+```
+
+A compiler should implement this in native code than using a regular expression engine.
+
+Examples of valid numeric literals:
+
+- `10` (an integer)
+- `512_000.000` (an integer)
+- `63_2.42` (a float)
+- `.425_000` (a float)
+- `50.2e+3` (an integer)
+- `4_2.1_2e8_2` (an integer)
+
+Examples of invalid numeric literals:
+
+- `10.` (this dot may be seen as a property access)
+- `1.5.1` (potentially a property access)
+- `4e-4.0` (potentially a property access)
+- `_52` (an identifier)
+- `10_` (parse error: numeric separator must be between two numbers)
+- `1__3` (parse error: numeric separator must be between two numbers)
+
+## Type of numeric literals
+
+Numeric literals do not have explicit types when in expressions, but rather adapt to the runtime-known types. This works exactly how [Zig's `comptime_int` type](https://ziglang.org/documentation/master/#toc-Runtime-Integer-Values) works (the type of integer literals), but this RFC does not give this type an observable name.
+
+```
+var x: u32 = 512
+// `512` becomes a u32 based on the type of `x`
+```
+
+Literals that do not fit in the result type will cause a compile error.
+
+```
+var x: u32 = -2
+// compile error: cannot coerce `-2` to `u32`
+```
+
+You cannot use [unary `~`](#integer-arithmetic) on an integer literal, it must be given a type, such as:
+
+```blight
+~512 // compile error: cannot infer type of constant numeric literal
+     // hint: unary operator '~' needs a result type
+
+~512.to(u32) // ok
+```
+
+Similarly, runtime-mutable variables cannot be assigned without annotating their type.
+
+> [!NOTE]
+> There is not yet an RFC for variable declarations. We will assume `var name: type` is the syntax.
+
+```blight
+var x = 512 // compile error: cannot infer type of constant numeric literal
+
+var x: u32 = 512 // ok
+var x = 512.to(u32) // ok, less preferred approach
+```
+
+## Constant Expression Evaluation
+
+Expressions that contain only numeric literals must be evaluated at compile time.
+
+```blight
+var x: u32 = (2.5 + 2.5) / 5
```

Comment:
Now I'm confused, in the compiler are these floats, bigints, or...?

Reply:
they are numeric literals. they do not have a type until assigned into a runtime-known variable.